In [2]:
# Data manipulation and handling libraries
import numpy as np
import pandas as pd

# Data visualization libraries
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as sci

# Data preprocessing and EDL Libraries
from collections import OrderedDict
from sklearn.preprocessing import StandardScaler

# Loading the Dataframe
df=pd.read_excel('Capstone Project.xlsx')

In [6]:
df.head()

cement   slag    ash  water  superplastic  coarseagg  fineagg  age  \
0   141.3  212.0    0.0  203.5           0.0      971.8    748.5   28   
1   168.9   42.2  124.3  158.3          10.8     1080.8    796.2   14   
2   250.0    0.0   95.7  187.4           5.5      956.9    861.2   28   
3   266.0  114.0    0.0  228.0           0.0      932.0    670.0   28   
4   154.8  183.4    0.0  193.3           9.1     1047.4    696.7   28   

   strength  
0     29.89  
1     23.51  
2     29.22  
3     45.85  
4     18.29

#Exploratory Data Analysis (EDA)

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1030 entries, 0 to 1029
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   cement        1030 non-null   float64
 1   slag          1030 non-null   float64
 2   ash           1030 non-null   float64
 3   water         1030 non-null   float64
 4   superplastic  1030 non-null   float64
 5   coarseagg     1030 non-null   float64
 6   fineagg       1030 non-null   float64
 7   age           1030 non-null   int64  
 8   strength      1030 non-null   float64
dtypes: float64(8), int64(1)
memory usage: 72.5 KB


Observations from df.info()
1) There are no null values
2) We have 1030 rows and 9 columns
3) Dependent variable = Strength
4) Rest of all columns are independent variable

In [9]:
df.describe()

cement         slag          ash        water  superplastic  \
count  1030.000000  1030.000000  1030.000000  1030.000000   1030.000000   
mean    281.167864    73.895825    54.188350   181.567282      6.204660   
std     104.506364    86.279342    63.997004    21.354219      5.973841   
min     102.000000     0.000000     0.000000   121.800000      0.000000   
25%     192.375000     0.000000     0.000000   164.900000      0.000000   
50%     272.900000    22.000000     0.000000   185.000000      6.400000   
75%     350.000000   142.950000   118.300000   192.000000     10.200000   
max     540.000000   359.400000   200.100000   247.000000     32.200000   

         coarseagg      fineagg          age     strength  
count  1030.000000  1030.000000  1030.000000  1030.000000  
mean    972.918932   773.580485    45.662136    35.817961  
std      77.753954    80.175980    63.169912    16.705742  
min     801.000000   594.000000     1.000000     2.330000  
25%     932.000000   730.950000     7.000000    23.710000  
50%     968.000000   779.500000    28.000000    34.445000  
75%    1029.400000   824.000000    56.000000    46.135000  
max    1145.000000   992.600000   365.000000    82.600000

Analysis from descriptive statistics
1) There might be skewness in the data in all columns
2) There might be chances of outliers if we compare quartile of some columns
3) Since minimum and q1 values are same for slag and ash, so we do not have outliers in the lower whisker region
4) The range of the value 'age' is from 1 to 365. We can say that the values are in 'days'


In [36]:
def custom_summary(my_df):
    result = []
    for col in my_df.columns:
        if my_df[col].dtypes != 'object':
            stats = OrderedDict({
                'feature_name':col,
                'count':my_df[col].count(),
                'quartile1':my_df[col].quantile(.25),
                'quartile2':my_df[col].quantile(.50),
                'quartile3':my_df[col].quantile(.75),
                'mean':my_df[col].mean(),
                'max':my_df[col].max(),
                'variance':round(my_df[col].var()),
                'standard_deviation':my_df[col].std(),
                'skewness':my_df[col].skew(),
                'kurtosis':my_df[col].kurt() 
            })

        result.append(stats)
    result_df=pd.DataFrame(result)
    
    #skewness type
    skewness_label =[]
    for i in result_df['skewness']:
        if i <= -1:
            skewness_label.append('Highly negatively skewed')
        elif -1 < i <= -0.5:
            skewness_label.append('Moderately negatively skewed')
        elif -0.5 < i < 0:
            skewness_label.append('Fairly negatively skewed')
        elif 0 <= i <= 1:
            skewness_label.append('Fairly Positively skewed')
        elif 0.5 <= i < 1:
            skewness_label.append('Moderately Positively skewed')
        elif i >= 1:
            skewness_label.append('Highly Positively skewed')
    result_df['skewness_comment'] = skewness_label
    
    #kurtosis type:
    kurtosis_label = []
    for i in result_df['kurtosis']:
        if i >= 1:
            kurtosis_label.append('Leptokurtic Curve')
        elif i <= 1:
            kurtosis_label.append('Platykurtic Curve')
        else:
            kurtosis_label.append('Mesokurtic Curve')
    
    result_df['kurtosis_comment'] = kurtosis_label
    
    #outliers
    outliers_label = []
    for col in my_df.columns:
        if my_df[col].dtypes != 'object':
            q1 = my_df[col].quantile(.25)
            q2 = my_df[col].quantile(.50)
            q3 = my_df[col].quantile(.75)
            iqr = q3 - q1
            lower_whisker = q1-0.25*iqr
            upper_whisker = q3+0.25*iqr
            if len(my_df[(my_df[col] < lower_whisker) | (my_df[col] > upper_whisker) ]) > 0:
                outliers_label.append('Have outliers')
            else:
                outliers_label.append('No outliers')
    result_df['Outlier Comment'] = outliers_label
    return result_df

In [37]:
custom_summary(df)

feature_name  count  quartile1  quartile2  quartile3        mean     max  \
0        cement   1030    192.375    272.900    350.000  281.167864   540.0   
1          slag   1030      0.000     22.000    142.950   73.895825   359.4   
2           ash   1030      0.000      0.000    118.300   54.188350   200.1   
3         water   1030    164.900    185.000    192.000  181.567282   247.0   
4  superplastic   1030      0.000      6.400     10.200    6.204660    32.2   
5     coarseagg   1030    932.000    968.000   1029.400  972.918932  1145.0   
6       fineagg   1030    730.950    779.500    824.000  773.580485   992.6   
7           age   1030      7.000     28.000     56.000   45.662136   365.0   
8      strength   1030     23.710     34.445     46.135   35.817961    82.6   

   variance  standard_deviation  skewness   kurtosis  \
0     10922          104.506364  0.509481  -0.520652   
1      7444           86.279342  0.800717  -0.508175   
2      4096           63.997004  0.537354  -1.328746   
3       456           21.354219  0.074628   0.122082   
4        36            5.973841  0.907203   1.411269   
5      6046           77.753954 -0.040220  -0.599016   
6      6428           80.175980 -0.253010  -0.102177   
7      3990           63.169912  3.269177  12.168989   
8       279           16.705742  0.416977  -0.313725   

           skewness_comment   kurtosis_comment Outlier Comment  
0  Fairly Positively skewed  Platykurtic Curve   Have outliers  
1  Fairly Positively skewed  Platykurtic Curve   Have outliers  
2  Fairly Positively skewed  Platykurtic Curve   Have outliers  
3  Fairly Positively skewed  Platykurtic Curve   Have outliers  
4  Fairly Positively skewed  Leptokurtic Curve   Have outliers  
5  Fairly negatively skewed  Platykurtic Curve   Have outliers  
6  Fairly negatively skewed  Platykurtic Curve   Have outliers  
7  Highly Positively skewed  Leptokurtic Curve   Have outliers  
8  Fairly Positively skewed  Platykurtic Curve   Have outliers